cli/src/parser.sk

```diff
@@ -150,61 +150,63 @@ fun parseArgsFrom(cmd: Command, argv: Sequence<String>): ParseResults {
 private fun updateValue(
   values: mutable Map<String, Value>,
   arg: Arg,
-  value: String,
-): ?ArgumentError {
+  value: Value,
+): Result<void, ArgumentError> {
   arg match {
   | ArrayArg _ ->
+    strValue = value match {
+    | StringValue(v) -> v
+    | _ -> invariant_violation(`Unexpected value type for ArrayArg`)
+    };
+
     values.maybeGet(arg.name) match {
     | Some(ArrayValue(arr)) ->
-      values.set(arg.name, ArrayValue(arr.concat(Array[value])))
-    | None() -> values.set(arg.name, ArrayValue(Array[value]))
+      values.set(arg.name, ArrayValue(arr.concat(Array[strValue])))
+    | None() -> values.set(arg.name, ArrayValue(Array[strValue]))
     | Some(v) -> invariant_violation(`Unexpected value type ${v} for ArrayArg`)
     }
-  | StringArg _ ->
-    if (!values.containsKey(arg.name)) {
-      values.set(arg.name, StringValue(value))
-    } else {
-      return Some(DuplicateValueError(arg.name))
-    }
-  | IntArg _ ->
+  | _ ->
     if (!values.containsKey(arg.name)) {
-      value.toIntOption() match {
-      | Some(v) -> values.set(arg.name, IntValue(v))
-      | _ -> return Some(InvalidArgumentError(arg.name))
-      }
+      values.set(arg.name, value)
     } else {
-      return Some(DuplicateValueError(arg.name))
+      return Failure(DuplicateValueError(arg.name))
     }
-  | _ -> invariant_violation(`Unexpected argument type ${inspect(arg)}`)
   };
 
-  None()
+  Success(void)
 }
 
-private fun parse(
-  cmd: Command,
-  iterator: mutable Iterator<String>,
-  values: mutable Map<String, Value> = mutable Map<String, Value>[],
-  globals: Array<Arg> = Array[],
-): ParseResults {
-  subcommands = mutable Map[];
-  for (subcmd in cmd._subcommands) {
-    if (subcommands.containsKey(subcmd.name)) {
+private fun defaultValue(arg: Arg): Value {
+  arg match {
+  | ValuedArg _ -> arg._default.default(MissingValue())
+  | BoolArg _ -> arg._default.default(BoolValue(false))
```

Comment:
Just a passing comment, but I don't understand why BoolArg is special in this way. Doesn't it make sense for bool args that default to true to be possible?

Reply:
They can indeed default to `true`, if `arg._default` is `Some(true)`. As for special casing the default value in the absence of an explicit default value, it is so that `args.getBool("foo")` returns `false` if `--foo` was not passed (and `foo` was declared as `BoolArg("foo")`.